eng/scripts/live-test-resource-cleanup.ps1

```diff
@@ -3,93 +3,373 @@
-Connect-AzAccount -Force -Tenant $TenantId -Credential $provisionerCredential -ServicePrincipal -Environment $Environment
-Select-AzSubscription -Subscription $SubscriptionId
+  $OwnerAliasCache[$Alias] = $false
 
-Write-Verbose "Fetching groups"
-$allGroups = @(Get-AzResourceGroup)
+  return $false;
+}
 
-Write-Host "Total Resource Groups: $($allGroups.Count)"
+function AddGithubUsersToAliasCache() {
+  if ($GithubAliasCachePath -and (Test-Path $GithubAliasCachePath)) {
+    Write-Host "Loading github -> microsoft alias mappings from filesystem cache."
+    $users = Get-Content $GithubAliasCachePath | ConvertFrom-Json -AsHashtable
+  } else {
+    Write-Host "Retrieving github -> microsoft alias mappings from opensource API."
+    $users = GetAllGithubUsers $OpensourceApiApplicationTenantId $OpensourceApiApplicationId $OpensourceApiApplicationSecret
+  }
+  if (!$users) {
+    Write-Error "Failed to retrieve github -> microsoft alias mappings from opensource api."
+    exit 1
+  }
+  foreach ($user in $users) {
+    if ($user.aad.alias) {
+      $OwnerAliasCache[$user.aad.alias] = $true
+    }
+  }
+  Write-Host "Found $($OwnerAliasCache.Count) github+microsoft user records."
+  if ($GithubAliasCachePath -and !(Test-Path $GithubAliasCachePath)) {
+      Write-Host "Caching github -> microsoft alias mappings to '$GithubAliasCachePath'"
+      $users | ConvertTo-Json -Depth 100 | Out-File $GithubAliasCachePath -WhatIf:$false
+  }
+}
 
-$now = [DateTime]::UtcNow
+function GetTag([object]$ResourceGroup, [string]$Key) {
+  if (!$ResourceGroup.Tags) {
+    return $null
+  }
 
-$noDeleteAfter = $allGroups.Where({ $_.Tags.Keys -notcontains "DeleteAfter" })
-Write-Host "Subscription contains $($noDeleteAfter.Count) resource groups with no DeleteAfter tags"
-$noDeleteAfter | ForEach-Object { Write-Verbose $_.ResourceGroupName }
+  foreach ($tagKey in $ResourceGroup.Tags.Keys) {
+    # Compare case-insensitive
+    if ($tagKey -ieq $Key) {
+      return $ResourceGroup.Tags[$tagKey]
+    }
+  }
 
-$hasDeleteAfter = $allGroups.Where({ $_.Tags.Keys -contains "DeleteAfter" })
-Write-Host "Count $($hasDeleteAfter.Count)"
-$toDelete = $hasDeleteAfter.Where({ $deleteDate = ($_.Tags.DeleteAfter -as [DateTime]); (!$deleteDate -or $now -gt $deleteDate) })
-Write-Host "Groups to delete: $($toDelete.Count)"
+  return $null
+}
 
-# Get purgeable resources already in a deleted state.
-$purgeableResources = @(Get-PurgeableResources)
+function HasValidOwnerTag([object]$ResourceGroup) {
+  $ownerTag = GetTag $ResourceGroup "Owners"
+  if (!$ownerTag) {
+    return $false
+  }
+  $owners = $ownerTag -split "[;, ]"
+  $hasValidOwner = $false
+  $invalidOwners = @()
+  foreach ($owner in $owners) {
+    if (IsValidAlias -Alias $owner) {
+      $hasValidOwner = $true
+    } else {
+      $invalidOwners += $owner
+    }
+  }
+  if ($invalidOwners) {
+    Write-Warning " Resource group '$($ResourceGroup.ResourceGroupName)' has invalid owner tags: $($invalidOwners -join ',')"
```

Comment:
We might want to consider removing invalid owners.

Reply:
Mentioned in chat: it does have the risk of a bug down the road where delete all the owners tags and then subsequently delete the group. Plus I'd be curious to see what the invalid owners are (e.g. if it was a typo, or a former employee, so that we can follow-up appropriately).

Reply:
agreed.